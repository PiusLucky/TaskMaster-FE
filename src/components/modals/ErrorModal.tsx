'use client';

import { ATOMS } from '@/api/atoms';
import { axiosErrorDefault } from '@/api/jotaiDefaultValues';
import { Dialog, DialogContent } from '@/components/ui/dialog';
import { useAtom } from 'jotai';

export function ErrorModal() {
  const [error, setError] = useAtom(ATOMS.axiosError);
  return (
    <Dialog
      open={error.errorOpenState}
      onOpenChange={() => setError(axiosErrorDefault)}
    >
      <DialogContent className='sm:max-w-[40rem] animate-in slide-in-from-bottom-48 !backdrop-blur-none'>
        <div className='flex flex-col items-center mb-8'>
          <div>
            <img
              src='/images/exclamation_circle.png'
              alt='Eclamation check icon'
              className='w-[5rem] md:w-[9rem]'
            />
          </div>
          <p className='text-[#BD8728] font-bold text-[1.2rem] md:text-[2rem]'>
            Error ({error.statusCode})
          </p>
          <p className='mt-2 text-center'>{error.errorMessage}</p>
        </div>
      </DialogContent>
    </Dialog>
  );
}
