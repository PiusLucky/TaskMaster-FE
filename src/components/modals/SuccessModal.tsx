'use client';

import { ATOMS } from '@/api/atoms';
import { axiosSuccessDefault } from '@/api/jotaiDefaultValues';
import { Dialog, DialogContent } from '@/components/ui/dialog';
import { useAtom } from 'jotai';

export function SuccessModal() {
  const [success, setSuccess] = useAtom(ATOMS.axiosSuccess);
  return (
    <Dialog
      open={success.successOpenState}
      onOpenChange={() => setSuccess(axiosSuccessDefault)}
    >
      <DialogContent className='sm:max-w-[40rem] animate-in slide-in-from-bottom-48 !backdrop-blur-none'>
        <div className='flex flex-col items-center mb-8'>
          <div>
            <img
              src='/images/circle_check.png'
              alt='Success check icon'
              className='w-[5rem] md:w-[9rem]'
            />
          </div>
          <p className='text-primary font-bold text-[1.2rem] md:text-[2rem]'>
            Success
          </p>
          <p className='mt-2 text-center'>{success.successMessage}</p>
        </div>
      </DialogContent>
    </Dialog>
  );
}
