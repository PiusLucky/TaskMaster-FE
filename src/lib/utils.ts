import { type ClassValue, clsx } from "clsx"
import { twMerge } from "tailwind-merge"
 
export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function capitalizeFirstLetter(inputString: string): string {
  if (inputString?.length === 0) {
    return inputString;
  }

  const firstLetter = inputString?.[0].toUpperCase();
  const restOfString = inputString?.slice(1);

  return firstLetter + restOfString;
}

export function getFirstLetters(inputString: string): string {
  const words: string[] = inputString.split(' ');
  const firstLetters: string = words.map((word) => word.charAt(0).toUpperCase()).join('')
  return firstLetters;
}

export function getCapitalizedFirstLetter(inputString: string): string {
  if (inputString?.length === 0) {
    return inputString;
  }

  const firstLetter = inputString?.charAt(0)?.toUpperCase();
  return firstLetter;
}

export function ellipisifyString(input: string, maxLength: number): string {
  if (input.length <= maxLength) {
    return input;
  }

  const startLength = Math.floor((maxLength - 3) / 2);
  const endLength = maxLength - 3 - startLength;

  const startPart = input.slice(0, startLength);
  const endPart = input.slice(input.length - endLength);

  return `${startPart}${endPart}...`;
}

export const objectToUri = (obj: { [key: string]: any }): string => {
  const uriString = Object.keys(obj)
    .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(obj[key])}`)
    .join('&');
  return uriString;
};