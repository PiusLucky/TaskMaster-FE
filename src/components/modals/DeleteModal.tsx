import { Dialog, DialogContent, DialogTrigger } from "@/components/ui/dialog";
import TMButton from "../common/TMButton";
import { useState } from "react";
import { apiClient } from "@/api";
import apiResources from "@/api/resources";
import { ATOMS } from "@/api/atoms";
import { useSetAtom } from "jotai";
import { STATUS_OK } from "@/lib/defaultConfig";

interface IProps {
  title: string;
  taskId: string;
  setIsPopOverOpen: (isPopOverOpen: boolean) => void;
  rerenderFunc: () => void;
}

function DeleteModal({
  title,
  taskId,
  setIsPopOverOpen,
  rerenderFunc,
}: IProps) {
  const [open, setOpen] = useState(false);
  const [loading, setIsLoading] = useState(false);
  const setError = useSetAtom(ATOMS.axiosError);
  const setSuccess = useSetAtom(ATOMS.axiosSuccess);
  const toggleOpenModal = () => {
    setOpen(!open);
  };

  const handleDeleteTask = async () => {
    setIsPopOverOpen(false);
    setIsLoading(true);

    try {
      const {
        meta: { statusCode },
      } = await apiClient.delete(
        apiResources.task,
        `/tasks/${taskId}`,
        setError,
        setSuccess
      );

      if (statusCode === STATUS_OK) {
        rerenderFunc();
      }
      setIsLoading(false);
    } catch (err) {
      setIsLoading(false);
    }
  };

  return (
    <Dialog open={open} onOpenChange={() => toggleOpenModal()}>
      <DialogTrigger asChild>
        <div
          className="flex gap-[0.4rem] cursor-pointer text-danger"
          onClick={() => toggleOpenModal()}
        >
          <img src="/svgs/trash.svg" alt="Pencil icon" />
          <p>Delete</p>
        </div>
      </DialogTrigger>
      <DialogContent className="sm:max-w-[60rem] h-[300px] md:max-h-full overflow-y-auto pt-8">
        <div className="pt-8">
          <div className="font-bold text-center mb-4">
            Are you sure you want to delete this task?
          </div>
          <div className="text-gray-500 mb-8 text-center">{title}</div>
          <div className="flex justify-center gap-[2.19rem]">
            <TMButton
              text="Cancel"
              classes="bg-white hover:bg-white text-black border border-black w-[9rem] !h-[3rem]"
              action={toggleOpenModal}
            />
            <TMButton
              text="Delete Task"
              classes="bg-danger hover:bg-danger text-white w-[9rem] !h-[3rem]"
              isLoading={loading}
              dataLoadingText="Wait..."
              action={handleDeleteTask}
            />
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
}

export default DeleteModal;
