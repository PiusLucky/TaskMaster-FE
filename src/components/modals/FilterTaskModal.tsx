import { Dialog, DialogContent, DialogTrigger } from "@/components/ui/dialog";
import { useCallback, useEffect, useState } from "react";
import { apiClient } from "@/api";
import apiResources from "@/api/resources";
import { useSetAtom } from "jotai";
import { ATOMS } from "@/api/atoms";
import TMButton from "../common/TMButton";
import * as z from "zod";

import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";

import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { CategoryType, PriorityType } from "@/lib/enum";

import { format } from "date-fns";
import { CalendarIcon } from "lucide-react";
import { cn, objectToUri } from "@/lib/utils";
import { Button } from "@/components/ui/button";
import { Calendar } from "@/components/ui/calendar";
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import { ITaskList } from "@/types/global-types";
import { TMEmitter } from "@/lib/eventEmitter";
import { FILTER_EVENT } from "@/lib/events";

function formatDate(inputDateString: string) {
  const inputDate = new Date(inputDateString);
  const year = inputDate.getFullYear();
  const month = (inputDate.getMonth() + 1).toString().padStart(2, "0"); // Adding 1 because months are 0-based
  const day = inputDate.getDate().toString().padStart(2, "0");

  return `${year}-${month}-${day}`;
}

const formSchema = z.object({
  category: z.string().optional(),
  priority: z.string().optional(),
  dueDate: z.date().optional(),
});

//Function starts here...
export function FilterTaskModal() {
  const [isCalendarOpen, setIsCalendarOpen] = useState(false);
  const [open, setOpen] = useState(false);
  const [actionButtonLoading, setActionButtonLoading] = useState(false);
  const setError = useSetAtom(ATOMS.axiosError);
  const toggleOpenModal = () => {
    setOpen(!open);
  };

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      category: "",
      priority: "",
    },
  });

  async function onSubmit(values: z.infer<typeof formSchema>) {
    setActionButtonLoading(true);
    try {
      const { category, priority, dueDate } = values;

      const uriQuery = {
        ...(category && {
          category: category ? category : undefined,
        }),
        ...(priority && {
          priority: priority ? priority : undefined,
        }),

        ...(dueDate && {
          dueDate: dueDate ? formatDate(dueDate?.toString()) : undefined,
        }),
      };

      const taskList = await apiClient.get<ITaskList>(
        apiResources.task,
        `/tasks?${objectToUri(uriQuery)}`,
        setError,
      );

      if (taskList) {
        TMEmitter.emit(FILTER_EVENT, { taskList });
        postActionCleanup();
      }

      setActionButtonLoading(false);
    } catch (err) {
      setActionButtonLoading(false);
    }
  }

  const postActionCleanup = () => {
    //Close modal
    toggleOpenModal();

    // Reset input
    form.reset();
  };

  return (
    <Dialog open={open} onOpenChange={() => toggleOpenModal()}>
      <DialogTrigger asChild>
        <TMButton
          text={"Filter"}
          iconRoute="/svgs/filter.svg"
          classes="w-[8.06rem] shadow-none !h-[3.37rem] rounded-[0.625rem] bg-white border border-[#000] text-[#000] hover:bg-white"
          action={() => toggleOpenModal()}
        />
      </DialogTrigger>
      <DialogContent className="sm:max-w-[60rem] max-h-[500px] md:max-h-full overflow-y-auto pt-8">
        <div>
          <p className="text-[#333] text-[1rem] font-[600] mb-[1.2rem]">
            Filter Tasks
          </p>
          <section>
            <Form {...form}>
              <form
                onSubmit={form.handleSubmit(onSubmit)}
                className="space-y-8"
              >
                {/* Category */}
                <FormField
                  control={form.control}
                  name="category"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Category</FormLabel>
                      <FormControl>
                        <Select
                          onValueChange={field.onChange}
                          defaultValue={field.value}
                        >
                          <SelectTrigger className="focus:!ring-primary">
                            <SelectValue placeholder="Select Category" />
                          </SelectTrigger>

                          <SelectContent>
                            <SelectGroup>
                              <SelectItem value={CategoryType.Work}>
                                Work
                              </SelectItem>
                              <SelectItem value={CategoryType.Studies}>
                                Studies
                              </SelectItem>
                              <SelectItem value={CategoryType.Fun}>
                                Fun
                              </SelectItem>
                              <SelectItem value={CategoryType.Other}>
                                Other
                              </SelectItem>
                            </SelectGroup>
                          </SelectContent>
                        </Select>
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />

                {/* Priority  & DueDate*/}
                <div className="flex md:items-center md:justify-between w-full gap-8 md:gap-16 flex-col md:flex-row">
                  <div className="flex-grow">
                    <FormField
                      control={form.control}
                      name="priority"
                      render={({ field }) => (
                        <FormItem>
                          <FormLabel>Priority</FormLabel>
                          <FormControl>
                            <Select
                              onValueChange={field.onChange}
                              defaultValue={field.value}
                            >
                              <SelectTrigger className="focus:!ring-primary">
                                <SelectValue placeholder="Select Priority" />
                              </SelectTrigger>

                              <SelectContent>
                                <SelectGroup>
                                  <SelectItem value={PriorityType.High}>
                                    High
                                  </SelectItem>
                                  <SelectItem value={PriorityType.Medium}>
                                    Medium
                                  </SelectItem>
                                  <SelectItem value={PriorityType.Low}>
                                    Low
                                  </SelectItem>
                                </SelectGroup>
                              </SelectContent>
                            </Select>
                          </FormControl>
                          <FormMessage />
                        </FormItem>
                      )}
                    />
                  </div>

                  <div className="flex-grow mt-2">
                    <FormField
                      control={form.control}
                      name="dueDate"
                      render={({ field }) => (
                        <FormItem className="flex flex-col">
                          <FormLabel>Due Date</FormLabel>
                          <Popover
                            open={isCalendarOpen}
                            onOpenChange={setIsCalendarOpen}
                          >
                            <PopoverTrigger asChild>
                              <FormControl>
                                <Button
                                  variant={"outline"}
                                  className={cn(
                                    "pl-3 text-left font-normal",
                                    !field.value && "text-muted-foreground",
                                  )}
                                >
                                  {field.value ? (
                                    format(field.value, "PPP")
                                  ) : (
                                    <span>Pick a date</span>
                                  )}
                                  <CalendarIcon className="ml-auto h-4 w-4 opacity-50" />
                                </Button>
                              </FormControl>
                            </PopoverTrigger>
                            <PopoverContent
                              className="w-auto p-0"
                              align="start"
                            >
                              <Calendar
                                mode="single"
                                selected={field.value}
                                onSelect={(e) => {
                                  field.onChange(e);
                                  setIsCalendarOpen(false);
                                }}
                                initialFocus
                              />
                            </PopoverContent>
                          </Popover>
                          <FormMessage />
                        </FormItem>
                      )}
                    />
                  </div>
                </div>

                <TMButton
                  text={"Apply filter"}
                  isSubmitable
                  width="w-full"
                  isLoading={actionButtonLoading}
                />
              </form>
            </Form>
          </section>
        </div>
      </DialogContent>
    </Dialog>
  );
}
