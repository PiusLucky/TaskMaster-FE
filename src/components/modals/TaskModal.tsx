import { Dialog, DialogContent, DialogTrigger } from "@/components/ui/dialog";
import { useCallback, useEffect, useState } from "react";
import { apiClient } from "@/api";
import apiResources from "@/api/resources";
import { useSetAtom } from "jotai";
import { ATOMS } from "@/api/atoms";
import TMButton from "../common/TMButton";
import * as z from "zod";
import { Textarea } from "@/components/ui/textarea";

import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";

import { Input } from "@/components/ui/input";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { CategoryType, PriorityType } from "@/lib/enum";

import { format } from "date-fns";
import { CalendarIcon } from "lucide-react";
import { cn } from "@/lib/utils";
import { Button } from "@/components/ui/button";
import { Calendar } from "@/components/ui/calendar";
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import { Task } from "@/types/global-types";

function formatDate(inputDateString: string) {
  const inputDate = new Date(inputDateString);
  const year = inputDate.getFullYear();
  const month = (inputDate.getMonth() + 1).toString().padStart(2, "0"); // Adding 1 because months are 0-based
  const day = inputDate.getDate().toString().padStart(2, "0");

  return `${year}-${month}-${day}`;
}

const formSchema = z.object({
  title: z
    .string()
    .min(5, {
      message: "Title must be at least 5 characters",
    })
    .max(255, {
      message: "Title must be at most 255 characters",
    }),
  description: z.string().min(5, {
    message: "Description must be at least 10 characters",
  }),
  category: z.string().min(1, {
    message: "Category is required",
  }),
  priority: z.string().min(1, {
    message: "Priority is required",
  }),
  dueDate: z.date({
    required_error: "Due date is required.",
  }),
});

export function TaskModel({
  rerenderParentComp,
  isCreate,
  task,
}: {
  rerenderParentComp: () => void;
  isCreate: boolean;
  task?: Task;
}) {
  const [isCalendarOpen, setIsCalendarOpen] = useState(false);
  const [open, setOpen] = useState(false);
  const [actionButtonLoading, setActionButtonLoading] = useState(false);
  const setError = useSetAtom(ATOMS.axiosError);
  const setSuccess = useSetAtom(ATOMS.axiosSuccess);
  const toggleOpenModal = () => {
    setOpen(!open);
  };

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      title: task?.title || "",
      description: task?.description || "",
      category: task?.category || "",
      priority: task?.priority || "",
      dueDate: task?.dueDate ? new Date(task?.dueDate) : new Date(),
    },
  });

  async function onSubmit(values: z.infer<typeof formSchema>) {
    setActionButtonLoading(true);
    try {
      const { title, description, category, priority, dueDate } = values;

      if (isCreate) {
        const { data } = await apiClient.post(
          apiResources.task,
          "/tasks",
          {
            title,
            description,
            category,
            priority,
            dueDate: formatDate(dueDate.toString()),
          },
          setError,
          setSuccess,
        );

        if (data) {
          postActionCleanup();
        }
      } else {
        const { data } = await apiClient.put(
          apiResources.task,
          `/tasks/${task?.id}`,
          {
            title,
            description,
            category,
            priority,
            dueDate: formatDate(dueDate.toString()),
          },
          setError,
          setSuccess,
        );

        if (data) {
          postActionCleanup();
        }
      }

      setActionButtonLoading(false);
    } catch (err) {
      setActionButtonLoading(false);
    }
  }

  const postActionCleanup = () => {
    // Refetch/Refresh parent component
    rerenderParentComp();

    //Close modal
    toggleOpenModal();

    // Reset input
    form.reset();
  };

  return (
    <Dialog open={open} onOpenChange={() => toggleOpenModal()}>
      <DialogTrigger asChild>
        {isCreate ? (
          <TMButton
            text={"Add New Task"}
            classes="w-[9.125rem] shadow-none !h-[3.37rem] rounded-[0.625rem]"
            action={() => toggleOpenModal()}
          />
        ) : (
          <div
            className="flex gap-[0.4rem] cursor-pointer"
            onClick={() => toggleOpenModal()}
          >
            <img src="/svgs/pencil.svg" alt="Pencil icon" />
            <p>Edit</p>
          </div>
        )}
      </DialogTrigger>
      <DialogContent className="sm:max-w-[60rem] max-h-[500px] md:max-h-full overflow-y-auto pt-8">
        <div>
          <p className="text-[#333] text-[1rem] font-[600] mb-[1.2rem]">
            {isCreate ? "Create" : "Update"} New Task
          </p>
          <section>
            <Form {...form}>
              <form
                onSubmit={form.handleSubmit(onSubmit)}
                className="space-y-8"
              >
                {/* Title */}
                <FormField
                  control={form.control}
                  name="title"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Title</FormLabel>
                      <FormControl className="relative">
                        <div>
                          <Input
                            placeholder="Enter Title"
                            {...field}
                            className=" focus:!ring-primary h-[3.125rem]"
                          />
                        </div>
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />

                {/* Category */}
                <FormField
                  control={form.control}
                  name="category"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Category</FormLabel>
                      <FormControl>
                        <Select
                          onValueChange={field.onChange}
                          defaultValue={field.value}
                        >
                          <SelectTrigger className="focus:!ring-primary">
                            <SelectValue placeholder="Select Category" />
                          </SelectTrigger>

                          <SelectContent>
                            <SelectGroup>
                              <SelectItem value={CategoryType.Work}>
                                Work
                              </SelectItem>
                              <SelectItem value={CategoryType.Studies}>
                                Studies
                              </SelectItem>
                              <SelectItem value={CategoryType.Fun}>
                                Fun
                              </SelectItem>
                              <SelectItem value={CategoryType.Other}>
                                Other
                              </SelectItem>
                            </SelectGroup>
                          </SelectContent>
                        </Select>
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />

                {/* Priority  & DueDate*/}
                <div className="flex md:items-center md:justify-between w-full gap-8 md:gap-16 flex-col md:flex-row">
                  <div className="flex-grow">
                    <FormField
                      control={form.control}
                      name="priority"
                      render={({ field }) => (
                        <FormItem>
                          <FormLabel>Priority</FormLabel>
                          <FormControl>
                            <Select
                              onValueChange={field.onChange}
                              defaultValue={field.value}
                            >
                              <SelectTrigger className="focus:!ring-primary">
                                <SelectValue placeholder="Select Priority" />
                              </SelectTrigger>

                              <SelectContent>
                                <SelectGroup>
                                  <SelectItem value={PriorityType.High}>
                                    High
                                  </SelectItem>
                                  <SelectItem value={PriorityType.Medium}>
                                    Medium
                                  </SelectItem>
                                  <SelectItem value={PriorityType.Low}>
                                    Low
                                  </SelectItem>
                                </SelectGroup>
                              </SelectContent>
                            </Select>
                          </FormControl>
                          <FormMessage />
                        </FormItem>
                      )}
                    />
                  </div>

                  <div className="flex-grow mt-2">
                    <FormField
                      control={form.control}
                      name="dueDate"
                      render={({ field }) => (
                        <FormItem className="flex flex-col">
                          <FormLabel>Due Date</FormLabel>
                          <Popover
                            open={isCalendarOpen}
                            onOpenChange={setIsCalendarOpen}
                          >
                            <PopoverTrigger asChild>
                              <FormControl>
                                <Button
                                  variant={"outline"}
                                  className={cn(
                                    "pl-3 text-left font-normal",
                                    !field.value && "text-muted-foreground",
                                  )}
                                >
                                  {field.value ? (
                                    format(field.value, "PPP")
                                  ) : (
                                    <span>Pick a date</span>
                                  )}
                                  <CalendarIcon className="ml-auto h-4 w-4 opacity-50" />
                                </Button>
                              </FormControl>
                            </PopoverTrigger>
                            <PopoverContent
                              className="w-auto p-0"
                              align="start"
                            >
                              <Calendar
                                mode="single"
                                selected={field.value}
                                onSelect={(e) => {
                                  field.onChange(e);
                                  setIsCalendarOpen(false);
                                }}
                                disabled={(date) =>
                                  date <=
                                  new Date(
                                    new Date().setDate(
                                      new Date().getDate() - 1,
                                    ),
                                  )
                                }
                                initialFocus
                              />
                            </PopoverContent>
                          </Popover>
                          <FormMessage />
                        </FormItem>
                      )}
                    />
                  </div>
                </div>

                {/* Description */}
                <div>
                  <FormField
                    control={form.control}
                    name="description"
                    render={({ field }) => (
                      <FormItem>
                        <FormLabel>Description</FormLabel>
                        <FormControl>
                          <div>
                            <Textarea
                              {...field}
                              placeholder="Enter task description"
                              className="focus:!ring-primary"
                            />
                          </div>
                        </FormControl>
                        <FormMessage />
                      </FormItem>
                    )}
                  />
                </div>

                <TMButton
                  text={isCreate ? "Create Task" : "Update Task"}
                  isSubmitable
                  width="w-full"
                  isLoading={actionButtonLoading}
                />
              </form>
            </Form>
          </section>
        </div>
      </DialogContent>
    </Dialog>
  );
}
