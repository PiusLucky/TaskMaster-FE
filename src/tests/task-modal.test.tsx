import { expect, test } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { TaskModal } from "@/components/modals/TaskModal";
import { hasInputValue } from "./utils";
import userEvent from "@testing-library/user-event";
import { SuccessModal } from "@/components/modals/SuccessModal";

const rerenderParentComp = () => {};

class MockPointerEvent extends Event {
  button: number;
  ctrlKey: boolean;
  pointerType: string;

  constructor(type: string, props: PointerEventInit) {
    super(type, props);
    this.button = props.button || 0;
    this.ctrlKey = props.ctrlKey || false;
    this.pointerType = props.pointerType || "mouse";
  }
}
window.PointerEvent = MockPointerEvent as any;
window.HTMLElement.prototype.scrollIntoView = vi.fn();
window.HTMLElement.prototype.releasePointerCapture = vi.fn();
window.HTMLElement.prototype.hasPointerCapture = vi.fn();

test("ADD TASK - user should be able to see the add task modal & make api request when add task buton is clicked", async () => {
  const user = userEvent.setup();
  const {
    queryByTestId: modalQueryByTestId,
    getByTestId,
    findByTestId,
  } = render(<TaskModal isCreate rerenderParentComp={rerenderParentComp} />);
  render(<SuccessModal />);

  //The modal should not show up when the Add New Task button has not been clicked
  const modalContent = modalQueryByTestId("modal-content");
  expect(modalContent).to.not.exist;
  const addNewTaskButton = getByTestId("task-modal-trigger");

  // Click the button
  fireEvent.click(addNewTaskButton);

  // Wait for the modal to appear by using findByTestId
  const createTaskDialog = await findByTestId("modal-content");
  expect(createTaskDialog).to.exist;

  const titleValue = getByTestId("title-input");
  const description = getByTestId("description");

  const testTitle = "New Task";
  const testDescription = "Test description";

  // Set all required inputs
  fireEvent.change(titleValue, { target: { value: testTitle } });
  fireEvent.change(description, { target: { value: testDescription } });

  // Find the button that triggers the dropdown
  const triggerCategory = screen.getByRole("combobox", {
    name: "Category",
  });

  const triggerPriority = screen.getByRole("combobox", {
    name: "Priority",
  });

  await user.click(triggerCategory);
  await user.click(screen.getByRole("option", { name: "Fun" })); //Set Fun as category

  await user.click(triggerPriority);
  await user.click(screen.getByRole("option", { name: "High" })); //Set High as priority

  expect(hasInputValue(titleValue, testTitle)).toBe(true);
  expect(hasInputValue(description, testDescription)).toBe(true);

  const submitButton = getByTestId("submit");
  expect(submitButton).to.exist;

  submitButton.click();

  // You should be able to see the success modal after submission
  const successTexts = await screen.findAllByText("Success");
  expect(successTexts.length).to.be.greaterThan(0);
});



test("UPDATE TASK - user should be able to see the update task modal & make api request when the pencil icon is clicked", async () => {
    const user = userEvent.setup();
    const {
      queryByTestId: modalQueryByTestId,
      getByTestId,
      findByTestId,
    } = render(<TaskModal isCreate={false} rerenderParentComp={rerenderParentComp} />);
    render(<SuccessModal />);
  
    //The modal should not show up when the Edit button has not been clicked
    const modalContent = modalQueryByTestId("modal-content");
    expect(modalContent).to.not.exist;
    const updateTaskButton = getByTestId("update-task-modal-trigger");
  
    // Click the button
    fireEvent.click(updateTaskButton);
  
    // Wait for the modal to appear by using findByTestId
    const createTaskDialog = await findByTestId("modal-content");
    expect(createTaskDialog).to.exist;
  
    // Should be able to create task
    const titleValue = getByTestId("title-input");
    const description = getByTestId("description");
  
    const testTitle = "New Task";
    const testDescription = "Test description";
  
    // Set all required inputs
    fireEvent.change(titleValue, { target: { value: testTitle } });
    fireEvent.change(description, { target: { value: testDescription } });
  
    const triggerCategory = screen.getByRole("combobox", {
      name: "Category",
    });
  
    const triggerPriority = screen.getByRole("combobox", {
      name: "Priority",
    });
  
    await user.click(triggerCategory);
    await user.click(screen.getByRole("option", { name: "Fun" })); //Set Fun as category
  
    await user.click(triggerPriority);
    await user.click(screen.getByRole("option", { name: "High" })); //Set High as priority
  
    expect(hasInputValue(titleValue, testTitle)).toBe(true);
    expect(hasInputValue(description, testDescription)).toBe(true);
  
    const submitButton = getByTestId("submit");
    expect(submitButton).to.exist;
  
    submitButton.click();
  
    // You should be able to see the success modal after submission
    const successTexts = await screen.findAllByText("Success");
    expect(successTexts.length).to.be.greaterThan(0);
  });

it("Should match snapshots", () => {
  const tree = renderer
    .create(<TaskModal isCreate rerenderParentComp={rerenderParentComp} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
