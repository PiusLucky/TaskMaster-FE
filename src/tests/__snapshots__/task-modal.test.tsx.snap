// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshots 1`] = `
<button
  className="inline-flex items-center justify-center text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-primary/90 h-10 px-4 py-2 bg-primary text-white md:w-[245px] md:h-[56px] select-none w-[9.125rem] shadow-none !h-[3.37rem] rounded-[0.625rem]"
  data-testid="task-modal-trigger"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <div
    className="flex gap-2 items-center"
  >
    Add New Task
  </div>
</button>
`;
